### PR TITLE
Feature/mlm

### PR DESCRIPTION
**Pull request by bobbycxy**
1) Added: masked language modelling (mlm) dataloader - called `MLMDataloader` - that masks inputs (80% masked, 10% randomised, 10% untouched), 
2) updated the cross_entropy to ignore ignore_index=-1, because -1 is the masked token id we'll use in the target (y) labels.
3) updated the build_trainer.py to have `MLMDataloader` in the dataloader dictionary.
4) created a mlm specific config.yaml file.

### CODE REVIEW

**Comment by DylanASHillier** [comment]
@LeonGuertler this is a beautful example of how a PR should be --> list of changes

File: trainers/loss_fn.py [hunk 1]
```diff
@@ -75,7 +75,7 @@ def compute_perplexity(logits, y, char_lengths, mask=None):
     # flatten both
     logits = logits.view(-1, logits.size(-1))
     y = y.view(-1)
-    loss = torch.nn.functional.cross_entropy(logits, y, reduction="none")
+    loss = torch.nn.functional.cross_entropy(logits, y, reduction="none", ignore_index=-1)
```
Comment (DylanASHillier) [inline trainers/loss_fn.py:78]: we are 100% sure this has no impacts elsewhere? should be okay but..

File: configs/full_configs/mlm_baseline.yaml [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+      attn_type: generic
+      num_heads: 16
+      normalization: rms_norm
+      group_size: 4
+      bias: false
+      is_causal: false
+  embedder:
+    tokenizer_type: gpt2
+    embedding_model_type: generic
+    dataset_name: simple_en_wiki
```
Comment (DylanASHillier) [inline configs/full_configs/mlm_baseline.yaml:20]: change this to stlm

File: trainers/loss_fn.py (hunk 1 above)
Comment (bobbycxy) [reply trainers/loss_fn.py:78]: Yes. Ignore_index's default value is -100. So, changing it from -100 to -1 would only impact existing label (y) assignments of -1 or -100. 

1) I searched our code base for the assignment of -100, there was none. 
2) I searched our code base for the assignment of -1. Apart from the new `MLMDataLoader` using label[~mask]=-1, other assignments were used in arguments such as 'dim=-1', or just as index values.

Comment (bobbycxy) [reply trainers/loss_fn.py:78]: image

Comment (bobbycxy) [reply trainers/loss_fn.py:78]: image

File: configs/full_configs/mlm_baseline.yaml (hunk 2 above)
Comment (bobbycxy) [reply configs/full_configs/mlm_baseline.yaml:20]: sorry, can I clarify what should be changed to stlm?

Comment (DylanASHillier) [reply configs/full_configs/mlm_baseline.yaml:20]: the dataset_name, but its okay, this will all be reworked later

**Comment by DylanASHillier** [comment]
I propose closing this, as we decided not to do MLM for now... feel free to rework this if you like, should be pretty simple to update, and happy to merge in since it doesn't add much code
